src/state/tls.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""gateway-api-integrator resource definition."""
+
+import dataclasses
+from typing import Dict
+
+import ops
+from ops.jujuversion import JujuVersion
+from ops.model import Relation
+
+TLS_CERTIFICATES_INTEGRATION = "certificates"
+
+
+class TlsIntegrationMissingError(Exception):
+    """Exception raised when _situation_.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the TlsIntegrationMissingError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+@dataclasses.dataclass(frozen=True)
+class TLSInformation:
+    """A component of charm state containing resource definition for kubernetes secret.
+
+    Attrs:
+        tls_requirer_integration: The integration instance with a TLS provider.
+        tls_certs: A dict of hostname: certificate obtained from the relation.
+        tls_keys: A dict of hostname: private_key stored in juju secrets.
+    """
+
+    tls_requirer_integration: Relation
```

Comment:
This is a bit tricky, as one of the goals of charmstate is [`Code should not directly interact with the state provided by Juju`](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619).

In this case, an ops model (that is not inmutable) is in the state.

There is clearly a reason to have that information available (update the relation data, grant secrets...), but not sure if there could be better place to hide it.

Reply:
I see your point, we can potentially exclude this from charm state and call every method of the `tls_certificate` module with an additional argument (the relation instance) which will solve the mutability question. I'll move the discussion internally and let's discuss there.

Reply:
I'm marking this as resolved and move this discussion internally since this is linked to Greg's comment about using lib to access relation data